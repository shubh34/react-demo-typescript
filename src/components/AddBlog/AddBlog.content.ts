export type contentType = {
    addBlogHeader: string,
    addBlogButtonTitle:  string,
    titleLabel:  string,
    blogContentLabel: string,
    categoriesLabel: string,
}
const content : contentType = {
    addBlogHeader: 'Add New Blog',
    addBlogButtonTitle: ' Add Blog',
    titleLabel: 'Title',
    blogContentLabel: 'Blog Content',
    categoriesLabel: 'Categories',

}

export default content;