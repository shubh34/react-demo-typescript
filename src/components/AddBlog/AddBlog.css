.form-categories {
    padding-bottom: 16px;
}