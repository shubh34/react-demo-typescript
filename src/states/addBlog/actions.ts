
export const INIT_ADD_BLOG: string = "INIT_ADD_BLOG";
export const UPDATE_FORM_FIELD: string = "UPDATE_FORM_FIELD";
export const ADD_BLOG_CATEGORY: string = "ADD_BLOG_CATEGORY";
export const DELETE_BLOG_CATEGORY: string = "DELETE_BLOG_CATEGORY";

export const initAddBlogForm = () => {
    return {
        type: INIT_ADD_BLOG,
    }
}

export const updateFormField = (name: string, value: string) => {
    return {
        type: UPDATE_FORM_FIELD,
        field: {
            name,
            value,
        }
    }
}

export const addCategory = (category: string) => {
    return {
        type: ADD_BLOG_CATEGORY,
        category,
    }
}

export const deleteCategory = (category: string) => {
    return {
        type: DELETE_BLOG_CATEGORY,
        category,
    }
}


