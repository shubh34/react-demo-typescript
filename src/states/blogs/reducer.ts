import { IBlogsState } from "./types";
import { ADD_BLOG } from "./action";

const init: IBlogsState = {
    blogs: []
};


const reducer = (state = init, action: any) : IBlogsState => {
	switch (action.type) {
	case ADD_BLOG: 
		return {
			...state,
			blogs: [...state.blogs, action.blog],
		}
	default:
		console.log('state', state);
		return state;
	}
};
export default reducer;
