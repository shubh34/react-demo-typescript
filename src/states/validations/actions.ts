
export const INIT_VALIDATION: string = "INIT_VALIDATION";
export const UPDATE_VALIDATION: string = "UPDATE_VALIDATION";
export const RESET_VALIDATION: string = "RESET_VALIDATION";

export const initValidation = (name: string, isValid: boolean) => {
    return {
        type: INIT_VALIDATION,
        field: {
            name,
            isValid,
        }
    }
}

export const updateValidation = (name: string, isValid: boolean) => {
    return {
        type: UPDATE_VALIDATION,
        field: {
            name,
            isValid,
        }
    }
}
export const resetValidation = () => {
    return {
        type: RESET_VALIDATION,
    }
}


