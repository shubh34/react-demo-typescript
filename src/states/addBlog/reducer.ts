import { IAddBlogState } from './types';
import { INIT_ADD_BLOG, UPDATE_FORM_FIELD, ADD_BLOG_CATEGORY, DELETE_BLOG_CATEGORY } from './actions';


const addBlogInitialState: IAddBlogState = {
   title:  '',
   categories: [],
   content: '',
}



const reducer = (state = addBlogInitialState, action: any): IAddBlogState => {
	switch (action.type) {
    case INIT_ADD_BLOG : 
        return {
            ...state,
            ...addBlogInitialState,
        }
    case UPDATE_FORM_FIELD: 
        return {
            ...state,
            [action.field.name]: action.field.value,
        }
    case ADD_BLOG_CATEGORY: 
        return {
            ...state,
            categories: [...state.categories, action.category]
        }
    case DELETE_BLOG_CATEGORY: 
        return {
            ...state,
            categories: state.categories.filter(category => category !== action.category)
        }
	default:

		return state;
	}
};
export default reducer;
