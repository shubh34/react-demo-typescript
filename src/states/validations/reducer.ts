import { IValidationState } from './types';
import { RESET_VALIDATION, INIT_VALIDATION, UPDATE_VALIDATION } from './actions';

const initialValidation: IValidationState = {
    form: {
        title: {
            isValid :  false, 
        },
        categories: {
            isValid :  false, 
        },
        content: {
            isValid: false,
        }
    },
}



const reducer = (state = initialValidation, action: any): IValidationState => {
	switch (action.type) {
    case RESET_VALIDATION : 
        return {
            ...state,
            ...initialValidation,
        }
    case INIT_VALIDATION: 
        return {
            ...state,
            form: {
                ...state.form,
                [action.field.name]: {
                    isValid: action.field.isValid,
                }
            }
        }
    case UPDATE_VALIDATION: 
        return {
            ...state,
            form: {
                ...state.form,
                [action.field.name]: {
                    isValid: action.field.isValid,
                }
            }
        }
     
	default:
		return state;
	}
};
export default reducer;
