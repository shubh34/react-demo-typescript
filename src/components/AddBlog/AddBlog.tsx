import React from 'react';
import { connect } from 'react-redux';
import { Button, Form} from 'semantic-ui-react';
import SectionHeader from '../sharedComponents/SectionHeader/SectionHeader';
import { IRootState } from '../../store';
import { isFormValid } from '../../states/validations/selectors';
import {
	initValidation,
	updateValidation,
	resetValidation,
} from '../../states/validations/actions';
import Title from './Forms/Title';
import Categories from './Forms/Categories';
import BlogContent from './Forms/BlogContent';
import {
	initAddBlogForm,
	updateFormField,
	addCategory,
	deleteCategory,
} from '../../states/addBlog/actions';
import {
	getAddBlogTitle,
	getAddBlogCategories,
	getAddBlogContent,
} from '../../states/addBlog/selectors';
import { addBlog } from '../../states/blogs/action';
import { IBlog } from '../../states/blogs/types';
import { RouteComponentProps, RouteProps } from 'react-router';

const mapState = (state: IRootState) => ({
	isFormValid: isFormValid(state),
	title: getAddBlogTitle(state),
	categories: getAddBlogCategories(state),
	blogContent: getAddBlogContent(state),
});

const mapDispatch = (dispatch: any) => ({
	initValidation: (name: string, isValid: boolean) => dispatch(initValidation(name, isValid)),
	updateValidation: (name: string, isValid: boolean) => dispatch(updateValidation(name, isValid)),
	resetValidation: () => dispatch(resetValidation()),
	initAddBlogForm: () => dispatch(initAddBlogForm()),
	updateFormField: (name: string, value: string) => dispatch(updateFormField(name, value)),
	addCategory: (categoryName: string) => dispatch(addCategory(categoryName)),
	addBlog: (blog: IBlog) => dispatch(addBlog(blog)),
	deleteCategory: (categoryName: string) => dispatch(deleteCategory(categoryName)),
});

type ReduxType = ReturnType<typeof mapState> & ReturnType<typeof mapDispatch>;

export class AddBlog extends React.Component<ReduxType & RouteComponentProps> {
	constructor(props: any) {
		super(props);
		this.onFormFieldChange = this.onFormFieldChange.bind(this);
		this.onAddCategory = this.onAddCategory.bind(this);
		this.onDeleteCategory = this.onDeleteCategory.bind(this);
		this.handleAddBlog = this.handleAddBlog.bind(this);
	}
	componentWillUnmount() {
		this.props.initAddBlogForm();
		this.props.resetValidation();
	}
	onFormFieldChange(name: string, value: string) {
		this.props.updateFormField(name, value);
		this.props.updateValidation(name, !!(value && value.length));
	}
	onAddCategory(category: string) {
		this.props.addCategory(category);
		if (this.props.categories.length === 0) {
			this.props.updateValidation('categories', true);
		}
	}
	onDeleteCategory(category: string) {
		this.props.deleteCategory(category);
		this.props.updateValidation('categories', this.props.categories.length > 1);
	}
	handleAddBlog() {
		console.log(this.props);
		const { title, blogContent: content, categories, addBlog, history } = this.props;
		const blog = {
			title,
			content,
			categories,
		};
		addBlog(blog);
		history.push('/');
	}

	render() {
		const { isFormValid, title, categories, blogContent } = this.props;
		return (
			<div>
				<SectionHeader header='Add New Blog' />
				<Form>
					<Title title={title} onChange={this.onFormFieldChange} />
					<BlogContent blogContent={blogContent} onChange={this.onFormFieldChange} />
					<Categories
						categories={categories}
						addCategory={this.onAddCategory}
						deleteCategory={this.onDeleteCategory}
					/>
					<Button id="add-blog" type="button" onClick={this.handleAddBlog} disabled={!isFormValid}>
						Add Blog
					</Button>
				</Form>
			</div>
		);
	}
}
export default connect(
	mapState,
	mapDispatch
)(AddBlog);
