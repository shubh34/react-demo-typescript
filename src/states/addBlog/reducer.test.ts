import reducer from  './reducer';
import {  initAddBlogForm, updateFormField, addCategory, deleteCategory } from './actions';
import { addBlogDefaultState } from '../../../tests_helpers/AddBlog';

describe('Reducer validation', () => {
    it('should init add blog form', () => {
        const blog = {
            title: 'My Blog',
            content: 'My content',
            categories: ['str']
        }
       expect(reducer(blog, initAddBlogForm())).toMatchSnapshot();
    });
    it('should update form field', () => {
        expect(reducer(addBlogDefaultState, updateFormField('title', 'My Blogs'))).toMatchSnapshot();
     });
     it('should add category', () => {
        expect(reducer(addBlogDefaultState, addCategory('Travel'))).toMatchSnapshot();
     });
     it('should delete category', () => {
        const blog = {
            title: 'My Blog',
            content: 'My content',
            categories: ['str']
        }
        expect(reducer(blog, deleteCategory('str'))).toMatchSnapshot();
     });
})