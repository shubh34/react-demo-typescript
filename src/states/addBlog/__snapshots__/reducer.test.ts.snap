// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer validation should add category 1`] = `
Object {
  "categories": Array [
    "test",
    "Travel",
  ],
  "content": "",
  "title": "",
}
`;

exports[`Reducer validation should delete category 1`] = `
Object {
  "categories": Array [],
  "content": "My content",
  "title": "My Blog",
}
`;

exports[`Reducer validation should init add blog form 1`] = `
Object {
  "categories": Array [],
  "content": "",
  "title": "",
}
`;

exports[`Reducer validation should update form field 1`] = `
Object {
  "categories": Array [
    "test",
  ],
  "content": "",
  "title": "My Blogs",
}
`;
